### ISSUE TITLE
Crash of HLTTauRefProducer:TauRefProducer in wf 136.7803 and 136.7801

### ISSUE THREAD

**Issue by silviodonato** [cms-sw/cmssw#29448]
After having merged #28417, we are getting a segmentation fault in wf 136.7803 and 136.7801

```
Begin processing the 1st record. Run 297227, Event 53500, LumiSection 1 on stream 0 at 09-Apr-2020 15:53:33.206 CEST


A fatal system signal has occurred: segmentation violation
The following is the call stack containing the origin of the signal.

Thu Apr  9 15:53:34 CEST 2020
Thread 7 (Thread 0xffff6ea49cb0 (LWP 44766)):
#0  0x0000ffff8d7ca770 in syscall () from /lib64/libc.so.6
#1  0x0000ffff8b331aec in XrdSys::LinuxSemaphore::Wait (this=0x1622f3d0) at /home/cmsbld/jenkins_c/workspace/auto-builds/CMSSW_11_1_0_pre5-slc7_aarch64_gcc820/build/CMSSW_11_1_0_pre5-build/BUILD/slc7_aarch64_gcc820/external/xrootd/4.10.0-bcolbf/xrootd-4.10.0/src/./XrdSys/XrdSysLinuxSemaphore.hh:161
#2  XrdCl::SyncQueue&lt;XrdCl::JobManager::JobHelper&gt;::Get (this=&lt;optimized out&gt;) at /home/cmsbld/jenkins_c/workspace/auto-builds/CMSSW_11_1_0_pre5-slc7_aarch64_gcc820/build/CMSSW_11_1_0_pre5-build/BUILD/slc7_aarch64_gcc820/external/xrootd/4.10.0-bcolbf/xrootd-4.10.0/src/./XrdCl/XrdClSyncQueue.hh:67
#3  XrdCl::JobManager::RunJobs (this=0x16230260) at /home/cmsbld/jenkins_c/workspace/auto-builds/CMSSW_11_1_0_pre5-slc7_aarch64_gcc820/build/CMSSW_11_1_0_pre5-build/BUILD/slc7_aarch64_gcc820/external/xrootd/4.10.0-bcolbf/xrootd-4.10.0/src/XrdCl/XrdClJobManager.cc:146
#4  0x0000ffff8b331d2c in RunRunnerThread (arg=&lt;optimized out&gt;) at /home/cmsbld/jenkins_c/workspace/auto-builds/CMSSW_11_1_0_pre5-slc7_aarch64_gcc820/build/CMSSW_11_1_0_pre5-build/BUILD/slc7_aarch64_gcc820/external/xrootd/4.10.0-bcolbf/xrootd-4.10.0/src/XrdCl/XrdClJobManager.cc:33
#5  0x0000ffff8d887d38 in start_thread () from /lib64/libpthread.so.0
#6  0x0000ffff8d7cf680 in thread_start () from /lib64/libc.so.6
Thread 6 (Thread 0xffff6f259cb0 (LWP 44765)):
#0  0x0000ffff8d7ca770 in syscall () from /lib64/libc.so.6
#1  0x0000ffff8b331aec in XrdSys::LinuxSemaphore::Wait (this=0x1622f3d0) at /home/cmsbld/jenkins_c/workspace/auto-builds/CMSSW_11_1_0_pre5-slc7_aarch64_gcc820/build/CMSSW_11_1_0_pre5-build/BUILD/slc7_aarch64_gcc820/external/xrootd/4.10.0-bcolbf/xrootd-4.10.0/src/./XrdSys/XrdSysLinuxSemaphore.hh:161
#2  XrdCl::SyncQueue&lt;XrdCl::JobManager::JobHelper&gt;::Get (this=&lt;optimized out&gt;) at /home/cmsbld/jenkins_c/workspace/auto-builds/CMSSW_11_1_0_pre5-slc7_aarch64_gcc820/build/CMSSW_11_1_0_pre5-build/BUILD/slc7_aarch64_gcc820/external/xrootd/4.10.0-bcolbf/xrootd-4.10.0/src/./XrdCl/XrdClSyncQueue.hh:67
#3  XrdCl::JobManager::RunJobs (this=0x16230260) at /home/cmsbld/jenkins_c/workspace/auto-builds/CMSSW_11_1_0_pre5-slc7_aarch64_gcc820/build/CMSSW_11_1_0_pre5-build/BUILD/slc7_aarch64_gcc820/external/xrootd/4.10.0-bcolbf/xrootd-4.10.0/src/XrdCl/XrdClJobManager.cc:146
#8  0x0000ffff8f7b691c in edm::global::EDProducerBase::doEvent(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ActivityRegistry*, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-04-09-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#9  0x0000ffff8f6e27a8 in edm::WorkerT&lt;edm::global::EDProducerBase&gt;::implDo(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-04-09-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#10 0x0000ffff8f68f160 in decltype ({parm#1}()) edm::convertException::wrap&lt;bool edm::Worker::runModule&lt;edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt; &gt;(edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt;::MyPrincipal const&amp, edm::EventSetupImpl const&amp, edm::StreamID, edm::ParentContext const&amp, edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt;::Context const*)::{lambda()#1}&gt;(bool edm::Worker::runModule&lt;edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt; &gt;(edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt;::MyPrincipal const&amp, edm::EventSetupImpl const&amp, edm::StreamID, edm::ParentContext const&amp, edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt;::Context const*)::{lambda()#1}) () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-04-09-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#11 0x0000ffff8f68f360 in bool edm::Worker::runModule&lt;edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt; &gt;(edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt;::MyPrincipal const&amp, edm::EventSetupImpl const&amp, edm::StreamID, edm::ParentContext const&amp, edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt;::Context const*) () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-04-09-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#12 0x0000ffff8f690c50 in std::__exception_ptr::exception_ptr edm::Worker::runModuleAfterAsyncPrefetch&lt;edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt; &gt;(std::__exception_ptr::exception_ptr const*, edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt;::MyPrincipal const&amp, edm::EventSetupImpl const&amp, edm::StreamID, edm::ParentContext const&amp, edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt;::Context const*) () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-04-09-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#13 0x0000ffff8f691c70 in edm::Worker::RunModuleTask&lt;edm::OccurrenceTraits&lt;edm::EventPrincipal, (edm::BranchActionType)1&gt; &gt;::execute() () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-04-09-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#14 0x0000ffff8dd2e550 in tbb::internal::custom_scheduler&lt;tbb::internal::IntelSchedulerTraits&gt;::process_bypass_loop (this=this@entry=0xffff8cfa3200, context_guard=..., t=t@entry=0xffff6deb2340, isolation=isolation@entry=0) at ../../include/tbb/task.h:992
#15 0x0000ffff8dd2e7a4 in tbb::internal::custom_scheduler&lt;tbb::internal::IntelSchedulerTraits&gt;::local_wait_for_all (this=0xffff8cfa3200, parent=..., child=&lt;optimized out&gt;) at ../../include/tbb/task.h:992
#16 0x0000ffff8f72cd38 in edm::EventProcessor::processLumis(std::shared_ptr&lt;void&gt; const&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-04-09-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#17 0x0000ffff8f73509c in edm::EventProcessor::runToCompletion() () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-04-09-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#18 0x00000000004129c4 in main::{lambda()#1}::operator()() const ()
#19 0x0000000000410c9c in main ()

Current Modules:

Module: HLTTauRefProducer:TauRefProducer (crashed)
```

**Comment by silviodonato** [comment]
assign reconstruction

**Comment by silviodonato** [comment]
@swozniewski

**Comment by slava77** [comment]
> we are getting a segmentation fault in wf 136.7803 and 136.7801

for the future reports, please provide a link or a more relevant part of the stack trace (the current snippet does not include the crashing thread)

IIUC, one example is in 
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/raw/slc7_amd64_gcc820/CMSSW_11_1_ROOT6_X_2020-04-09-2300/pyRelValMatrixLogs/run/136.7801_RunHLTPhy2017B_AOD+RunHLTPhy2017B_AOD+DQMHLTonAOD_2017+HARVESTDQMHLTonAOD_2017/step2_RunHLTPhy2017B_AOD+RunHLTPhy2017B_AOD+DQMHLTonAOD_2017+HARVESTDQMHLTonAOD_2017.log

with the crashing thread (just look for sig_dostack_then_abort +5 lines
```
#3  0x00002b1cf69bb5a9 in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_ROOT6_X_2020-04-09-2300/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#4  &lt;signal handler called&gt;
#5  0x00002b1cec0e6ae4 in edm::parameterSet(edm::Provenance const&amp, edm::ProcessHistory const&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_ROOT6_X_2020-04-09-2300/lib/slc7_amd64_gcc820/libFWCoreCommon.so
#6  0x00002b1d3d2ee12b in HLTTauRefProducer::doPFTaus(edm::StreamID, edm::Event&amp) const () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_ROOT6_X_2020-04-09-2300/lib/slc7_amd64_gcc820/pluginDQMOfflineTriggerPlugins.so
#7  0x00002b1d3d2ef2cd in HLTTauRefProducer::produce(edm::StreamID, edm::Event&amp, edm::EventSetup const&amp) const () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_ROOT6_X_2020-04-09-2300/lib/slc7_amd64_gcc820/pluginDQMOfflineTriggerPlugins.so
#8  0x00002b1cec558ca9 in edm::global::EDProducerBase::doEvent(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ActivityRegistry*, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02623/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_ROOT6_X_2020-04-09-2300/lib/slc7_amd64_gcc820/libFWCoreFramework.so
```

**Comment by Dr15Jones** [comment]
@makortel FYI

**Comment by slava77** [comment]
I guess that the problem is here
https://github.com/cms-tau-pog/cmssw/blob/e3ee2f54d22194b79a0f8ad58d32cfb1b8b946e2/DQMOffline/Trigger/plugins/HLTTauRefProducer.cc#L147-L158

**Comment by slava77** [comment]
@swozniewski @mbluj 
please post a quick note if you are available to check/fix the problem today (or before Monday)

**Comment by makortel** [comment]
Right, the problem is that the product is missing (same problem as in #29447?), and the `edm::Handle::provenance()` returns a `nullptr` which is deferenced without checking for `nullptr`
https://github.com/cms-sw/cmssw/blob/3d03fbb3659b0a3e4bfdde43350c5923d0645b6c/DQMOffline/Trigger/plugins/HLTTauRefProducer.cc#L148-L149

**Comment by slava77** [comment]
> Right, the problem is that the product is missing (same problem as in #29447?)

OK, I guess for this file we are better off to use a throwing variant of ```Event::getHandle```
so that the error is reported properly.
Is there such a thing? I didn't find it in the interface

**Comment by makortel** [comment]
> > Right, the problem is that the product is missing (same problem as in #29447?)
> 
> OK, I guess for this file we are better off to use a throwing variant of `Event::getHandle`
> so that the error is reported properly.
> Is there such a thing? I didn't find it in the interface

There is `edm::ValidHandle` that can be created from `edm::Handle` with `edm::makeValid()`, that throws an exception if the `Handle` is invalid. Unfortunately the `ValidHandle` does not provide access to provenance (very likely because it is intended to be used with `edm::OrphanHandle` as well).

We could think of changing `edm::HandleBase::provenance()` to return a reference and throw an exception in absence of provenance object (and adding `provenancePtr()` to return the pointer without checks).

**Comment by mbluj** [comment]
On Fri, Apr 10, 2020 at 3:48 PM Slava Krutelyov <notifications@github.com>
wrote:

> @swozniewski <https://github.com/swozniewski> @mbluj
> <https://github.com/mbluj>
> please post a quick note if you are available to check/fix the problem
> today (or before Monday)
>
I am sorry, but I cannot take a look before next Tuesday,
BTW, is it a new problem? Can it be related with some recently integrated
stuff? Could it be somewhat related to the new tauID dataformat or it is
caused by something not directly related to tau reco&ID?


<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by mbluj** [comment]
> 
> @swozniewski @mbluj
> please post a quick note if you are available to check/fix the problem today (or before Monday)

I am sorry, but I cannot take a look before next Tuesday.

**Comment by slava77** [comment]
>  Could it be somewhat related to the new tauID dataformat

it is related to #28417

**Comment by swozniewski** [comment]
Sorry for the late reply. I was offline since friday. I can work on this tomorrow. In another producer I added these two lines https://github.com/cms-sw/cmssw/pull/28417/files#diff-c7c5c83df6323f1a53a05655cd079d3fR377-R378  in order to avoid segfaults when accessing provenance. The module would still crash due to a missing input at a later stage but in a better way than this segfault. So I guess it would be good to add this here as well unless you treat this at a more fundamental position as discussed above. In order to check why the workflow is failing now, could you please let me know how to run it? It is not part of the usual matrix test, right, otherwise we would have detected the problem earlier?

**Comment by slava77** [comment]
> these two lines https://github.com/cms-sw/cmssw/pull/28417/files#diff-c7c5c83df6323f1a53a05655cd079d3fR377-R378

My browser didn't point to anything given this link/anchor tag.



> In order to check why the workflow is failing now, could you please let me know how to run it?

The issue description mentions 136.7803 
```runTheMatrix.py -l 136.7803 ```

**Comment by silviodonato** [comment]
After #29453 the error is more readable
```
----- Begin Fatal Exception 13-Apr-2020 01:07:43 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 297557 lumi: 123 event: 196868183 stream: 2
   [1] Running path 'dqmoffline_step'
   [2] Calling method for module HLTTauRefProducer/'TauRefProducer'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::ValueMap&lt;reco::SingleTauDiscriminatorContainer&gt;
Looking for module label: hpsPFTauBasicDiscriminators
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc820/CMSSW_11_1_X_2020-04-12-2300/pyRelValMatrixLogs/run/136.7803_RunHLTPhy2017B_RAWAOD+RunHLTPhy2017B_RAWAOD+DQMHLTonRAWAOD_2017+HARVESTDQMHLTonAOD_2017/step2_RunHLTPhy2017B_RAWAOD+RunHLTPhy2017B_RAWAOD+DQMHLTonRAWAOD_2017+HARVESTDQMHLTonAOD_2017.log#/

**Comment by swozniewski** [comment]
> > these two lines https://github.com/cms-sw/cmssw/pull/28417/files#diff-c7c5c83df6323f1a53a05655cd079d3fR377-R378
> 
> My browser didn't point to anything given this link/anchor tag.

https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATTauProducer.cc#L377-L378

**Comment by swozniewski** [comment]
But ok, you already added #29453 . Thanks!

**Comment by swozniewski** [comment]
I think that the problem is that this workflow loads existing files with the old data format. Therefore, we would need an era modifier here that applies the old config: https://github.com/cms-sw/cmssw/blob/master/DQMOffline/Trigger/python/HLTTauDQMOffline_cfi.py#L9-L21
The workflow is using Run2_2017. I don't see a suitable modifier here: https://github.com/cms-sw/cmssw/blob/master/Configuration/Eras/python/Era_Run2_2017_cff.py Should we add another one?

**Comment by makortel** [comment]
The segfault was fixed in https://github.com/cms-sw/cmssw/pull/29453, so could the further discussion on the configuration be moved to https://github.com/cms-sw/cmssw/issues/29447, and we close this issue?

**Comment by slava77** [comment]
> The segfault was fixed in #29453, so could the further discussion on the configuration be moved to #29447, and we close this issue?

not really, this is not the same as  #29447 .
This is the case where either this issue should be renamed or a new one should be created.
For renaming, I could suggest "Failure running HLT DQM on old AOD inputs in wf 136.7803 and 136.7801"

**Comment by slava77** [comment]
> The workflow is using Run2_2017. I don't see a suitable modifier here: https://github.com/cms-sw/cmssw/blob/master/Configuration/Eras/python/Era_Run2_2017_cff.py Should we add another one?

this will be a global problem for workflows like this, we need something like ```tau_ID_rerun``` modifier

**Comment by swozniewski** [comment]
Shall I provide separate PRs for this issue and #29447 or rather a common one?

**Comment by makortel** [comment]
> We could think of changing `edm::HandleBase::provenance()` to return a reference and throw an exception in absence of provenance object (and adding `provenancePtr()` to return the pointer without checks).

I created an issue https://github.com/cms-sw/cmssw/issues/29464 for this idea, it could make sense anyway despite of the quick fix of https://github.com/cms-sw/cmssw/pull/29453.

**Comment by slava77** [comment]
> Shall I provide separate PRs for this issue and #29447 or rather a common one?

IIUC, the fix in #29447 is not necessarily related to rerunning tauID, it's about broken tau skims.
So, a combined PR is not required. Also, a fix in #29447  is more urgent

**Comment by silviodonato** [comment]
Fixed by #29467
